### PR TITLE
Version 3: Implement ExUnit max-failures feature

### PR DESCRIPTION
**Pull request by eksperimental**
I decided to submit a new PR and keep the previous PR #8174

Implements: #4586
Related: #8174, #7480, #8129

/cc @ibizaman, @hylkealons, @Kartstig, @stevegraham, @DenisGorbachev, @fertapric

### CODE REVIEW

**Comment by eksperimental** [comment]
@josevalim I initially decided to go with the approach of loading all test cases, because I think it is important to inform the user how many tests are in total. But now that the implementation is fully working, if it needs to be removed it's easier to do if you considered so.

**Comment by eksperimental** [comment]
I'm running OTP 21.0.6, and the `make clean test` is passing, I don't know why CI is failing in every version

**Comment by josevalim** [comment]
Thanks @eksperimental! We are getting closed to have this merged but the PR is still mixed with unrelated changes. For example, renaming of variables, change of aliases. Can you please revert those changes? The runner code is complex so we should do the minimum amount of changes so we can focus on reviewing the important parts. The renaming and refactoring of related parts can be done in a separate PR.

**Comment by josevalim** [comment]
Also, to keep this simpler, let's also revert the statuses changes. I don't want to add a new status "not executed" or at least not in the initial implementation. So we shouldn't worry about tracking or notifying those. We can make this clear in the failure message "--max-failures reached, aborting test suite".

**Comment by eksperimental** [comment]
It will be less time consuming for me to submit a PR that renames variables and change the order of arguments in functions, and build thIS PR on top of that, and it will be quick to be reviewed and approve than this one. Ley me know if that suits you

**Comment by josevalim** [comment]
Your call. :)

**Comment by eksperimental** [comment]
Just to be clear, so a test with 10 tests, where 2nd, 3rd and 4th fail, and --max-failures 2. What should the stats return? Total: 3, failures: 2
Or should we return nil because the suites been aborted?

**Comment by josevalim** [comment]
As soon as we collect --max-failures, we will stop reporting the other tests and failures. We will just do enough work to make sure they terminate correctly.

**Comment by eksperimental** [comment]
This is finally ready for a review

File: lib/ex_unit/lib/ex_unit.ex [hunk 1]
```diff
@@ -62,11 +62,11 @@ defmodule ExUnit do
...
 
-    1. Passed (also represented by `nil`)
+    1. Passed (represented by `nil`)
     2. Failed
-    3. Skipped (via @tag :skip)
-    4. Excluded (via :exclude filters)
-    5. Invalid (when setup_all fails)
+    3. Skipped (via `@tag :skip`)
+    4. Excluded (via `:exclude` filters)
+    5. Invalid (when `setup_all/1,2` fail)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit.ex]: Unrelated to this change, please do it in a separate commit.

File: lib/ex_unit/lib/ex_unit/cli_formatter.ex [hunk 2]
```diff
@@ -153,8 +153,29 @@ defmodule ExUnit.CLIFormatter do
     test_counter =
       Enum.reduce(test_module.tests, config.test_counter, &update_test_counter(&2, &1))
 
-    failure_counter = config.failure_counter + tests_length
-    config = %{config | test_counter: test_counter, failure_counter: failure_counter}
+    config = %{
+      config
+      | test_counter: test_counter,
+        failure_counter: config.failure_counter + tests_length
+    }
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/cli_formatter.ex]: Unrelated to this change, please revert.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 3]
```diff
@@ -332,6 +416,87 @@ defmodule ExUnit.Runner do
...
+  end
+
+  defp register_test_module(pid) do
+    ExUnit.OnExitHandler.register(pid)
+  end
+
+  defp get_failure_counter(stats_pid) when is_pid(stats_pid),
+    do: RunnerStats.get_failure_counter(stats_pid)
+
+  defp increment_failure_counter(stats_pid, %struct{state: {tag, _}}, increment \\ 1)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Please avoid default arguments on private functions.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 4]
```diff
@@ -332,6 +416,87 @@ defmodule ExUnit.Runner do
...
+  defp process_failure(
+         process_failure_config,
+         %ExUnit.TestModule{state: {tag, _}, tests: tests} = test_module
+       )
+       when tag in [:failed, :invalid] do
+    %{config: config} = process_failure_config
+    failure_counter = increment_failure_counter(config.stats_pid, test_module, length(tests))
+
+    if max_failures_reached?(failure_counter, config.max_failures) do
+      max_failures_has_been_reached(config.manager, config.stats_pid, nil)
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Now that I see the code, I think killing all running tests will be harmful and potentially generate tons of logs. That's because if the test process has spawned other processes, killing the test process will be considered as a cascading failure. I think it is best just to not kill them and let them run their "course".

Also, don't we need to make a distinction between "max_failures" has just been reached and "max failures" has been surpassed? When it has just been reached, you still want to log that test but you want to stop the run. For subsequent cases, you don't want to log, nor continue the run.

File: lib/ex_unit/lib/ex_unit.ex (hunk 1 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit.ex]: done

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 3 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: done

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 4 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: Just wondering if exiting like this would avoid the logging and also the cascading effect
```elixir
Process.exit(registered_pid, {:shutdown, :exit_max_failures_reached})
```

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: So I have implemented your suggestion. Let me know if that works.
Regarding max failures reached and surpassed, we I don't think we need that, because I log the tests first and then I do `process_failures` which is based on the passed test, and just then I update the failure counter, if it's reached, I launch the `max_failures_have_been_reached` function, and any subsequent test will just be ignored and marked as :not_executed if they were still running

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 5]
```diff
@@ -107,28 +118,39 @@ defmodule ExUnit.Runner do
...
+      Enum.each(test_results.pending, fn test ->
+        EM.test_started(config.manager, test)
+        EM.test_finished(config.manager, test)
+      end)
+
+      test_module = %{test_module | tests: test_results.finished}
+      EM.module_finished(config.manager, test_module)
+
+      # message loop/2
+      send(config.runner_pid, {self(), :module_finished})
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: I see you simplified the message by not including the module, could this be further simplified by just sending to `loop/3` the messages `:module_finished` and `:max_failures_reached`?

Btw, the comment should say `loop/3` :) But, is it necessary?

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 6]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
 
-    {exec_on_exit(test_module, module_pid, timeout), pending, finished_tests}
+    test_module = exec_on_exit(test_module, module_pid, timeout)
+    {test_module, test_results}
+  end
+
+  defp spawn_module_monitor(config, test_module, parent_pid, tests) do
+    spawn_monitor(fn ->
+      module_pid = self()
+      register(module_pid)
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: We are missing information here: "on exit".

Comment (fertapric) [reply lib/ex_unit/lib/ex_unit/runner.ex]: In addition, I prefer `self()` because it always means "the current process"

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 7]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+      end
+
+      exit(:shutdown)
+    end)
+  end
+
+  # returns a map in the shape of
+  # %{:not_executed => list(), :finished => list()}
+  defp run_tests(config, tests, context) do
+    Enum.map(tests, fn test ->
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Minor nitpick:

```elixir
tests
|> Enum.map(tests, &run_test(config, &1, context)) # or  fn test -> run_test(config, test, context)
|> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
|> Map.put_new(:not_executed, [])
|> Map.put_new(:finished, [])
```

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 8]
```diff
@@ -107,28 +118,39 @@ defmodule ExUnit.Runner do
...
+      EM.module_started(config.manager, test_module)
+
+      # Prepare tests, selecting which ones should be run or skipped
+      tests = prepare_tests(config, test_module.tests)
+
+      {test_module, test_results} =
+        if Enum.all?(tests, & &1.state) do
+          # The pending tests here aren't actually run,
+          # so they're marked as "finished".
+          {test_module, %{pending: tests, finished: tests}}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Can you please revert those changes? Previously it was returning a tuple with three elements and it should continue to return a tuple with three elements, as this change is not related to this PR.

I have already said this a couple times but these unrelated changes make the PR really hard to review. After a couple rounds, we are still finding layers of code that were just moved around. Next time, please, please consider making the PR as small as possible. Any refactoring should happen elsewhere.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 5 above)
Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: Please let's continue sending the three element tuple. If the previous element was not necessary, it should be removed in another PR.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 8 above)
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: @eksperimental I like this change, but it generates some noise in this not-so-small PR. 

I understand that while working on big PRs like this one you'll see a lot of points for improvement, but please, try to keep you focused on the main goal of the PR 🙂

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 9]
```diff
@@ -316,13 +420,17 @@ defmodule ExUnit.Runner do
...
 
   defp exec_on_exit(test_or_case, pid, timeout) do
-    case ExUnit.OnExitHandler.run(pid, timeout) do
-      :ok ->
-        test_or_case
-
-      {kind, reason, stack} ->
-        state = test_or_case.state || failed(kind, reason, prune_stacktrace(stack))
-        %{test_or_case | state: state}
+    if ExUnit.OnExitHandler.registered?(pid) do
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Why do we need to wrap this on `registered?`? `run` should return `:ok` if there is nothing registered. We probably don't need a pre-check.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 10]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
-    {test_module, pending, finished_tests} =
+    {test_module, test_results} =
       receive do
-        {^module_pid, :module_finished, test_module, failed_tests, finished_tests} ->
-          Process.demonitor(module_ref, [:flush])
-          {test_module, failed_tests, finished_tests}
+        {^module_pid, :module_finished, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Let's continue with the three element tuple here too. No need to put it in a map only to extract it out.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 11]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+          else
+            process_failure(config, test_module)
+            Process.demonitor(module_ref, [:flush])
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: This call exists in both `if` and `else` branches, so we should probably move it up.

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 7 above)
Comment (josevalim) [reply lib/ex_unit/lib/ex_unit/runner.ex]: We should not return a map here though. In fact, we should stop running the other tests as soon as we get the first `not_executed`.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 12]
```diff
@@ -217,68 +286,101 @@ defmodule ExUnit.Runner do
...
-        run_test_with_capture_log(capture_log, config, test, Map.merge(tags, context))
-      else
-        test
+    if max_failures_reached?(config.stats_pid, config.max_failures) do
+      {:not_executed, test}
+    else
+      EM.test_started(config.manager, test)
+
+      # We need to retrieve the status, because the test can be aborted
+      # during its execution due to max_failures being reached
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: When could this happen?

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 13]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
-        end
-
-        exit(:shutdown)
-      end)
+    {module_pid, module_ref} = spawn_module_monitor(config, test_module, parent_pid, tests)
 
-    {test_module, pending, finished_tests} =
+    {test_module, test_results} =
       receive do
-        {^module_pid, :module_finished, test_module, failed_tests, finished_tests} ->
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Please let's keep this same message format as before.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 14]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
-          Process.demonitor(module_ref, [:flush])
-          {test_module, failed_tests, finished_tests}
+        {^module_pid, :module_finished, test_module, tests} ->
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            process_failure(config, test_module)
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Let's keep it the same three element tuple as before too.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 15]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            {test_module, %{pending: [], finished: []}}
+          else
+            test_module = %{test_module | state: failed({:EXIT, module_pid}, error, [])}
+            process_failure(config, test_module)
+            {test_module, %{pending: [], finished: []}}
+          end
       end
 
-    {exec_on_exit(test_module, module_pid, timeout), pending, finished_tests}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Let's keep it the same three element tuple as before too.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 16]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+             %{pending: [], finished: tests_run.finished}}
+          )
+
+        {:error, test_module} ->
+          invalid_tests = Enum.map(tests, &%{&1 | state: {:invalid, test_module}})
+
+          send(
+            parent_pid,
+            {module_pid, :module_finished, test_module, %{pending: invalid_tests, finished: []}}
+          )
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: We should not change the message format.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 17]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+      case exec_module_setup(test_module) do
+        {:ok, _test_module, context} ->
+          # max_failures can be reached during the execution of test_module,
+          # so we keep track of which tests are not executed and which ones are finished.
+          tests_run = run_tests(config, tests, context)
+
+          send(
+            parent_pid,
+            {module_pid, :module_finished, test_module,
+             %{pending: [], finished: tests_run.finished}}
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: We should not change the message format.

File: lib/mix/lib/mix/tasks/test.ex [hunk 18]
```diff
@@ -569,6 +574,25 @@ defmodule Mix.Tasks.Test do
     end
   end
 
+  defp normalize_max_failures(opts) do
+    case Keyword.fetch(opts, :max_failures) do
+      {:ok, value} when value in ["infinity", ":infinity"] ->
+        Keyword.put(opts, :max_failures, :infinity)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: I don't think we need to support this. The equivalent of setting it to infinity is not setting it at all.

File: lib/mix/test/mix/tasks/test_test.exs [hunk 19]
```diff
@@ -257,6 +258,75 @@ defmodule Mix.Tasks.TestTest do
     end)
   end
 
+  describe "--max-failures" do
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/test_test.exs]: I don't believe we need those integration tests as they are very expensive. We should be able to check all of those cases through the ExUnit suite itself.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 20]
```diff
@@ -217,68 +286,101 @@ defmodule ExUnit.Runner do
...
 
-  defp receive_test_reply(test, test_pid, test_ref, timeout) do
+  defp receive_test_reply(config, test, test_pid, test_ref, timeout) do
     receive do
       {^test_pid, :test_finished, test} ->
+        if max_failures_reached?(config.stats_pid, config.max_failures) do
+          {:not_executed, test}
+        else
+          process_failure(config, test)
+          Process.demonitor(test_ref, [:flush])
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: The demonitor should be moved up.

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 21]
```diff
@@ -157,35 +179,82 @@ defmodule ExUnit.Runner do
...
+          if max_failures_reached?(config.stats_pid, config.max_failures) do
+            Process.demonitor(module_ref, [:flush])
+            {test_module, %{pending: [], finished: []}}
+          else
+            process_failure(config, test_module)
+            Process.demonitor(module_ref, [:flush])
+            {test_module, tests}
+          end
+
+        {^module_pid, :max_failures_reached, test_module, tests} ->
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/runner.ex]: I don't see when this message would be ever delivered. Is it still necessary?

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 6 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: @fertapric what do you exactly mean by missing on exit information?

Comment (fertapric) [reply lib/ex_unit/lib/ex_unit/runner.ex]: The previous code was `ExUnit.OnExitHandler.register(pid)` and now is `register(pid)`. Before it was clear that you were registering this process for something related to “on exit” callbacks.

File: lib/mix/test/mix/tasks/test_test.exs [hunk 22]
```diff
@@ -257,6 +258,75 @@ defmodule Mix.Tasks.TestTest do
...
+
+        assert mix(:concurrent, ~w[test --max-failures 2 --max-cases 3]) =~
+                 "\n5 tests, 2 failures\n"
+
+        assert mix(:concurrent, ~w[test --max-failures 8 --max-cases 10]) =~
+                 "\n12 tests, 3 failures, 5 invalid\n"
+      end)
+    end
+
+    test "aborts concurrent running tests" do
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/test_test.exs]: This test and the other two below are excellent. Can we please move them to the ExUnit suite? They are really testing ExUnit and not the mix integration. Thanks! :heart:

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 6 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: True. It was a left over from when we were also registering the PIDs in the runnerstats

**Comment by josevalim** [comment]
Thank you! Merged manually in 8c373044dcde1654670d8da10448e4b8fecf0b93! 
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

Note I did some changes before merging. Namely, I reduced the amount of times we check for `max_reached_finished?`, as it felt a lot of times we should have known the answer as long as we passed it around.

I have also pushed some fixes that made it work properly for umbrellas.

**Comment by stevegraham** [comment]
Thank you @eksperimental! 🙌🏻
